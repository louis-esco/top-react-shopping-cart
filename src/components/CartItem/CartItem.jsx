import styles from "./CartItem.module.css";
import PropTypes from "prop-types";

export default function CartItem({ item, addToCart, removeFromCart }) {
  return (
    <div key={item.id} className={styles.item}>
      <div className={styles.itemImg}>
        <img src={item.image} alt="" />
      </div>
      <div className={styles.itemContent}>
        <div className={styles.itemTitle}>{item.title}</div>
        <div className="itemPrice">Unit price: {item.price}€</div>
        <div className={styles.itemQty}>
          Quantity:{" "}
          <button
            onClick={() => removeFromCart(item, 1)}
            className={styles.qtyBtn}
          >
            -
          </button>{" "}
          <div>{item.qty}</div>{" "}
          <button className={styles.qtyBtn} onClick={() => addToCart(item, 1)}>
            +
          </button>
        </div>
        <div className="itemTotal">Total price: {item.qty * item.price}€</div>
        <button
          className={styles.removeBtn}
          onClick={() => {
            removeFromCart(item, item.qty);
          }}
        >
          &#10005; Remove from cart
        </button>
      </div>
    </div>
  );
}

CartItem.propTypes = {
  item: PropTypes.object,
  addToCart: PropTypes.func,
  removeFromCart: PropTypes.func,
};
