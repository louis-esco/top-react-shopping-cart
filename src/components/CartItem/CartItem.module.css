.item {
  display: flex;
  gap: 10px;
  font-size: 0.9rem;
  font-weight: 300;
}

.itemImg {
  flex: 0 0 200px;
  height: 130px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.itemImg img {
  max-width: 100%;
  max-height: 100%;
}

.itemContent {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.itemTitle {
  font-weight: 600;
}

.itemQty {
  display: flex;
  gap: 5px;
}

.qtyBtn {
  width: 20px;
  aspect-ratio: 1;
  border-radius: 3px;
}

.qtyBtn:hover,
.removeBtn:hover {
  cursor: pointer;
}

.removeBtn,
.qtyBtn {
  background-color: #5b5b5b;
  border: none;
  color: #ffffff;
}

.removeBtn {
  width: fit-content;
  padding: 5px 10px;
  border-radius: 3px;
}
