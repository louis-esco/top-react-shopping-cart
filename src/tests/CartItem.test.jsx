import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CartItem from "../components/CartItem/CartItem";

describe("CartItem component", () => {
  const addToCart = vi.fn();
  const removeFromCart = vi.fn();
  const item = {
    id: 1,
    qty: 2,
    title: "Nice item",
    price: 64,
    description: "You can do lots with this item",
    category: "electronics",
    image: "https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg",
    rating: {
      rate: 3.3,
      count: 203,
    },
  };

  it("Render CartItem", () => {
    const { container } = render(
      <CartItem
        item={item}
        addToCart={addToCart}
        removeFromCart={removeFromCart}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("Increment and decrement buttons call addToCart and removeFromCart", async () => {
    const user = userEvent.setup();

    render(
      <CartItem
        item={item}
        addToCart={addToCart}
        removeFromCart={removeFromCart}
      />
    );
    const incrementBtn = screen.getByRole("button", { name: "+" });
    const decrementBtn = screen.getByRole("button", { name: "-" });

    await user.click(incrementBtn);

    expect(addToCart).toHaveBeenCalled();

    await user.click(decrementBtn);

    expect(removeFromCart).toHaveBeenCalled();
  });
});
