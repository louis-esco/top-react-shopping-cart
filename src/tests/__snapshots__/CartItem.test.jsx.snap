// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartItem component > Render CartItem 1`] = `
<div>
  <div
    class="_item_61ad6d"
  >
    <div
      class="_itemImg_61ad6d"
    >
      <img
        alt=""
        src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg"
      />
    </div>
    <div
      class="_itemContent_61ad6d"
    >
      <div
        class="_itemTitle_61ad6d"
      >
        Nice item
      </div>
      <div
        class="itemPrice"
      >
        Unit price: 
        64
        €
      </div>
      <div
        class="_itemQty_61ad6d"
      >
        Quantity:
         
        <button
          class="_qtyBtn_61ad6d"
        >
          -
        </button>
         
        <div>
          2
        </div>
         
        <button
          class="_qtyBtn_61ad6d"
        >
          +
        </button>
      </div>
      <div
        class="itemTotal"
      >
        Total price: 
        128
        €
      </div>
      <button
        class="_removeBtn_61ad6d"
      >
        ✕ Remove from cart
      </button>
    </div>
  </div>
</div>
`;
